WordPress/src/main/java/org/wordpress/android/ui/accounts/LoginActivity.java

```diff
@@ -1,60 +1,110 @@
 package org.wordpress.android.ui.accounts;
 
 import android.os.Bundle;
+import android.support.v4.app.Fragment;
 import android.support.v4.app.FragmentTransaction;
 import android.support.v7.app.AppCompatActivity;
+import android.view.MenuItem;
 import android.widget.Toast;
 
 import org.wordpress.android.R;
 import org.wordpress.android.WordPress;
-import org.wordpress.android.ui.accounts.login.LogInOrSignUpFragment;
-import org.wordpress.android.ui.accounts.login.nav.LoginEvents;
-import org.wordpress.android.ui.accounts.login.nav.LoginNavigationController;
-import org.wordpress.android.ui.accounts.login.nav.LoginNavigationFsmGetter;
-import org.wordpress.android.ui.accounts.login.nav.LoginState;
+import org.wordpress.android.ui.accounts.login.LoginEmailFragment;
+import org.wordpress.android.ui.accounts.login.LoginPrologueFragment;
+import org.wordpress.android.ui.accounts.login.LoginNavFragment;
+import org.wordpress.android.ui.accounts.login.nav.LoginNavHandler;
+import org.wordpress.android.ui.accounts.login.nav.LoginStateGetter;
 
-public class LoginActivity extends AppCompatActivity implements
-        LoginNavigationFsmGetter,
-        LoginNavigationController.ContextImplementation {
-
-    LoginNavigationController mLoginNavigationController = new LoginNavigationController(LoginState.PROLOGUE, this);
+public class LoginActivity extends AppCompatActivity implements LoginNavHandler, LoginStateGetter.FsmGetter {
+    private static final String TAG_LOGIN_NAV_FRAGMENT = "TAG_LOGIN_NAV_FRAGMENT";
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         ((WordPress) getApplication()).component().inject(this);
 
-        setContentView(R.layout.welcome_activity);
+        setContentView(R.layout.login_activity);
+
+        if (!hasLoginNavFragment()) {
+            addLoginNavFragment();
+        }
 
         if (savedInstanceState == null) {
             addLoginPrologueFragment();
         }
     }
 
-    protected void addLoginPrologueFragment() {
-        LogInOrSignUpFragment loginSignupFragment = new LogInOrSignUpFragment();
+    @Override
+    public boolean onOptionsItemSelected(final MenuItem item) {
+        if (item.getItemId() == android.R.id.home) {
+            onBackPressed();
+            return true;
+        }
+        return super.onOptionsItemSelected(item);
+    }
+
+    @Override
+    public void onBackPressed() {
+        // perform the Nav back first so updated fragments can find their expected state there.
+        getLoginNavFragment().goBack();
+
+        super.onBackPressed();
+    }
+
+    /**
+     * Retrieves the login navigation fragment from the FragmentManager.
+     *  **NOTE** Should not be used too early before the navigation fragment has been added to the FragmentManager.
```

Comment:
This comment got me thinking now that I have a better understanding of how things work. Do you think we should be making a check for it here? I am guessing this method is very important for us and things will just not work if this fails, so that'll mean we actually want it to fail so we can fix issues. But I am wondering, if the fail itself (crash logs) will be specific enough for us to understand that it's caused by a call to this method. Do you think it'd make sense to have a check here and throw a special exception, just so we know where to look?

P.S: Again, this is just thinking out loud, not really saying we should do something or another.